src/eval.ts

```diff
@@ -0,0 +1,341 @@
+'use strict';
+
+import {AstFactory} from './ast_factory';
+
+const _BINARY_OPERATORS = {
+  '+': function(a: any, b: any) {
+    return a + b;
+  },
+  '-': function(a: any, b: any) {
+    return a - b;
+  },
+  '*': function(a: any, b: any) {
+    return a * b;
+  },
+  '/': function(a: any, b: any) {
+    return a / b;
+  },
+  '%': function(a: any, b: any) {
+    return a % b;
+  },
+  '==': function(a: any, b: any) {
+    return a == b;  // tslint:disable-line: triple-equals
+  },
+  '!=': function(a: any, b: any) {
+    return a != b;  // tslint:disable-line: triple-equals
+  },
+  '===': function(a: any, b: any) {
+    return a === b;
+  },
+  '!==': function(a: any, b: any) {
+    return a !== b;
+  },
+  '>': function(a: any, b: any) {
+    return a > b;
+  },
+  '>=': function(a: any, b: any) {
+    return a >= b;
+  },
+  '<': function(a: any, b: any) {
+    return a < b;
+  },
+  '<=': function(a: any, b: any) {
+    return a <= b;
+  },
+  '||': function(a: any, b: any) {
+    return a || b;
+  },
+  '&&': function(a: any, b: any) {
+    return a && b;
+  },
+  '|': function(a: any, f: Function) {
+    return f(a);
+  },
+};
+
+const _UNARY_OPERATORS = {
+  '+': function(a: any) {
+    return a;
+  },
+  '-': function(a: any) {
+    return -a;
+  },
+  '!': function(a: any) {
+    return !a;
+  },
+};
+
+export interface Scope { [key: string]: any; }
+
+export interface BaseNode {
+  evaluate: (this: this, scope: Scope) => any;
+  getIds(this: this, idents: string[]): string[];
+}
+
+export type Node = Literal | Empty | ID | Unary | Binary | Getter | Invoke |
+    Index | Ternary | Map | List;
+export interface Literal extends BaseNode {
```

Comment:
If you import the expression interfaces from `'./ast'`, you can use intersection types to define these types:

```typescript
export type Literal = EvaluatableExpression & ast.Literal;
```